proof/refine/ARM/Syscall_R.thy

```diff
@@ -1636,26 +1636,425 @@ lemma setSchedulerAction_obj_at'[wp]:
+  for cte_wp_at[wp]: "cte_wp_at P c"
+
+lemma handle_timeout_valid_sched_action:
+  "\<lbrace>\<lambda>s. valid_sched_action s \<and> released_ipc_queues s \<and> scheduler_act_not tptr s
+      \<and> active_sc_valid_refills s
+      \<and> (is_timeout_fault ex \<and> active_sc_tcb_at tptr s \<or> released_if_bound_sc_tcb_at tptr s)\<rbrace>
+   handle_timeout tptr ex
+   \<lbrace>\<lambda>_. valid_sched_action :: det_state \<Rightarrow> _\<rbrace>"
+  unfolding handle_timeout_def
+  apply (wpsimp wp: send_fault_ipc_valid_sched_action)
+  done
+
+lemma end_timeslice_valid_sched_action:
+  "\<lbrace>valid_sched_action and released_ipc_queues and active_sc_valid_refills and scheduler_act_sane
+    and schact_is_rct and cur_sc_tcb and cur_sc_active and (\<lambda>s. sym_refs (state_refs_of s))\<rbrace>
+   end_timeslice canTimeout
+   \<lbrace>\<lambda>_. valid_sched_action :: det_state \<Rightarrow> _\<rbrace>"
+  unfolding end_timeslice_def
+  apply (wpsimp wp: handle_timeout_valid_sched_action postpone_valid_sched_action)
+  apply (intro conjI; clarsimp simp: is_timeout_fault_def cur_sc_tcb_def dest!: schact_is_rct)
+  apply (drule (1) sym_refs_bound_sc_tcb_iff_sc_tcb_sc_at[OF refl refl, THEN iffD2])
+  apply (clarsimp simp: active_sc_tcb_at_def2)
+  done
+
+lemma endTimeslice_invs'[wp]:
+  "\<lbrace>invs' and ct_active' and sch_act_sane\<rbrace>
+   endTimeslice timeout
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  unfolding endTimeslice_def
+  apply (wpsimp wp: handleTimeout_invs' isValidTimeoutHandler_inv hoare_drop_imp)
+  apply (clarsimp simp: runnable_eq_active')
+  apply (frule (1) active_ex_cap'[OF _ invs_iflive'])
+  apply (clarsimp simp: ct_in_state'_def sch_act_sane_def)
+  done
+
+crunches setConsumedTime, updateSchedContext
+  for sch_act_sane[wp]: sch_act_sane
+  and ct_active'[wp]: ct_active'
+  (simp: sch_act_sane_def ct_in_state'_def ignore: setSchedContext)
+
+crunches refillResetRR, refillBudgetCheck
+  for ct_active'[wp]: ct_active'
+  and sch_act_sane[wp]: sch_act_sane
+  and ex_nonz_cap_to'[wp]: "ex_nonz_cap_to' p"
+  and typ_at'[wp]: "\<lambda>s. Q (typ_at' P p s)"
+  (wp: crunch_wps)
+
+crunches chargeBudget
+  for typ_at'[wp]: "\<lambda>s. Q (typ_at' P p s)"
+  (wp: crunch_wps)
+
+lemma refillResetRR_invs'[wp]:
+  "\<lbrace>invs' and ct_active' and sch_act_sane\<rbrace>
+   refillResetRR scp
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  unfolding refillResetRR_def
+  apply (wpsimp wp: updateSchedContext_invs')
+  apply (intro conjI; clarsimp elim!: live_sc'_ex_cap[OF invs_iflive'])
+  by (fastforce dest!: valid_sc_strengthen[OF invs_valid_objs']
+                 simp: valid_sched_context'_def valid_sched_context_size'_def scBits_simps objBits_simps')
+
+lemma cur_sc_tcb_cross:
+  "\<lbrakk>(s, s') \<in> state_relation; valid_objs s; pspace_aligned s; pspace_distinct s;
+    cur_sc_tcb s; schact_is_rct s\<rbrakk>
+  \<Longrightarrow> obj_at' (\<lambda>sc. scTCB sc = Some (ksCurThread s')) (ksCurSc s') s'"
```

Comment:
This could probably be moved earlier as well, in case we want to cross it over more often.